File: src/modules/cart/cart.service.ts
import { Types } from 'mongoose';
import { AppError } from '../../shared/errors/AppError.js';
import { recordAudit } from '../../shared/utils/audit.js';
import { Product } from '../catalog/product.model.js';
import { Cart, MAX_ITEM_QUANTITY } from './cart.model.js';

// The cart stores only references and quantities. Every price shown here is
// re-read from the product document at read time, and the order module
// re-derives totals from the database at checkout — the backend never trusts
// client-supplied prices (a core security rule of this project).

export interface CartItemPublic {
  productId: string;
  variantId: string;
  sku: string;
  productName: string;
  productSlug: string;
  image: string | null;
  attributes: Record<string, string>;
  unitPriceCents: number;
  compareAtPriceCents: number | null;
  taxRate: number;
  quantity: number;
  available: number;
  lineTotalCents: number;
  inStock: boolean;
  productAvailable: boolean;
}

export interface CartPublic {
  items: CartItemPublic[];
  itemCount: number;
  subtotalCents: number;
}

async function getOrCreateCart(userId: string): Promise<InstanceType<typeof Cart>> {
  return Cart.findOneAndUpdate(
    { userId },
    { $setOnInsert: { items: [] } },
    { upsert: true, new: true },
  );
}

function findVariant(product: { variants: ReadonlyArray<unknown> }, variantId: string) {
  return product.variants.find(
    (v) => (v as { _id?: { toString(): string } })._id?.toString() === variantId,
  );
}

export async function getCart(userId: string): Promise<CartPublic> {
  const cart = await getOrCreateCart(userId);
  return enrichCart(cart.items);
}

async function enrichCart(
  items: Array<{ productId: Types.ObjectId; variantId: Types.ObjectId; quantity: number }>,
): Promise<CartPublic> {
  if (items.length === 0) {
    return { items: [], itemCount: 0, subtotalCents: 0 };
  }

  const variantIds = items.map((i) => i.variantId.toString());
  const products = await Product.find({ 'variants._id': { $in: variantIds } }).lean();

  const publicItems: CartItemPublic[] = [];
  let subtotalCents = 0;

  for (const item of items) {
    const product = products.find((p) =>
      (p.variants as unknown as Array<{ _id: unknown }>).some(
        (v) => (v._id as { toString(): string }).toString() === item.variantId.toString(),
      ),
    );
    const variant = product ? findVariant(product, item.variantId.toString()) : undefined;

    if (!product || !variant) {
      publicItems.push({
        productId: item.productId.toString(),
        variantId: item.variantId.toString(),
        sku: '',
        productName: 'Unavailable product',
        productSlug: '',
        image: null,
        attributes: {},
        unitPriceCents: 0,
        compareAtPriceCents: null,
        taxRate: 0,
        quantity: item.quantity,
        available: 0,
        lineTotalCents: 0,
        inStock: false,
        productAvailable: false,
      });
      continue;
    }

    const variantWithStock = variant as unknown as {
      sku: string;
      attributes: Record<string, string>;
      priceCents: number;
      compareAtPriceCents: number | null;
      taxRate: number;
      stock: { available: number };
      isActive: boolean;
      images: string[];
    };
    const unitPriceCents = variantWithStock.priceCents;
    const lineTotalCents = unitPriceCents * item.quantity;
    subtotalCents += lineTotalCents;

    publicItems.push({
      productId: product._id.toString(),
      variantId: item.variantId.toString(),
      sku: variantWithStock.sku,
      productName: product.name,
      productSlug: product.slug,
      image: product.images[0] ?? variantWithStock.images[0] ?? null,
      attributes: variantWithStock.attributes,
      unitPriceCents,
      compareAtPriceCents: variantWithStock.compareAtPriceCents,
      taxRate: variantWithStock.taxRate,
      quantity: item.quantity,
      available: variantWithStock.stock.available,
      lineTotalCents,
      inStock: variantWithStock.stock.available > 0,
      productAvailable:
        product.status === 'published' && product.isActive === true && variantWithStock.isActive,
    });
  }

  return {
    items: publicItems,
    itemCount: items.reduce((sum, i) => sum + i.quantity, 0),
    subtotalCents,
  };
}

async function resolveVariant(productId: string, variantId: string) {
  const product = await Product.findOne({ _id: productId, 'variants._id': variantId }).lean();
  if (!product)
    throw AppError.badRequest('Product variant does not exist', { errorCode: 'VARIANT_NOT_FOUND' });
  if (product.status !== 'published' || product.isActive !== true) {
    throw AppError.badRequest('This product is not available for purchase', {
      errorCode: 'PRODUCT_UNAVAILABLE',
    });
  }
  const variant = findVariant(product, variantId);
  if (!variant)
    throw AppError.badRequest('Product variant does not exist', { errorCode: 'VARIANT_NOT_FOUND' });
  const variantWithStock = variant as unknown as {
    isActive: boolean;
    stock: { available: number };
  };
  if (!variantWithStock.isActive) {
    throw AppError.badRequest('This variant is not available', {
      errorCode: 'VARIANT_UNAVAILABLE',
    });
  }
  return { product, variant: variantWithStock };
}

export async function addItem(
  userId: string,
  productId: string,
  variantId: string,
  quantity: number,
): Promise<CartPublic> {
  await resolveVariant(productId, variantId);

  const cart = await getOrCreateCart(userId);
  const existing = cart.items.find((i) => i.variantId.toString() === variantId);
  const newQuantity = (existing?.quantity ?? 0) + quantity;

  if (newQuantity > MAX_ITEM_QUANTITY) {
    throw AppError.badRequest(`Maximum of ${MAX_ITEM_QUANTITY} units per item`, {
      errorCode: 'QUANTITY_LIMIT_EXCEEDED',
    });
  }

  // Cart quantity can never exceed current sellable stock; the authoritative
  // guard is applied again atomically at checkout, but failing early gives the
  // customer feedback while browsing.
  const availability = await checkAvailability(productId, variantId);
  if (newQuantity > availability) {
    throw AppError.badRequest(`Only ${availability} units are currently available`, {
      errorCode: 'INSUFFICIENT_STOCK',
    });
  }

  if (existing) {
    existing.quantity = newQuantity;
  } else {
    cart.items.push({
      productId: new Types.ObjectId(productId),
      variantId: new Types.ObjectId(variantId),
      quantity,
      addedAt: new Date(),
    });
  }
  await cart.save();
  recordAudit('cart.item_added', { userId, productId, variantId, quantity });
  return enrichCart(cart.items);
}

export async function updateItemQuantity(
  userId: string,
  variantId: string,
  quantity: number,
): Promise<CartPublic> {
  const cart = await getOrCreateCart(userId);
  const item = cart.items.find((i) => i.variantId.toString() === variantId);
  if (!item) throw AppError.notFound('Item not found in cart');

  if (quantity > MAX_ITEM_QUANTITY) {
    throw AppError.badRequest(`Maximum of ${MAX_ITEM_QUANTITY} units per item`, {
      errorCode: 'QUANTITY_LIMIT_EXCEEDED',
    });
  }

  const availability = await checkAvailability(item.productId.toString(), variantId);
  if (quantity > availability) {
    throw AppError.badRequest(`Only ${availability} units are currently available`, {
      errorCode: 'INSUFFICIENT_STOCK',
    });
  }

  item.quantity = quantity;
  await cart.save();
  recordAudit('cart.item_updated', { userId, variantId, quantity });
  return enrichCart(cart.items);
}

export async function removeItem(userId: string, variantId: string): Promise<CartPublic> {
  const cart = await getOrCreateCart(userId);
  const before = cart.items.length;
  cart.items = cart.items.filter((i) => i.variantId.toString() !== variantId);
  if (cart.items.length === before) throw AppError.notFound('Item not found in cart');
  await cart.save();
  recordAudit('cart.item_removed', { userId, variantId });
  return enrichCart(cart.items);
}

export async function clearCart(userId: string): Promise<CartPublic> {
  const cart = await getOrCreateCart(userId);
  cart.items = [];
  await cart.save();
  recordAudit('cart.cleared', { userId });
  return { items: [], itemCount: 0, subtotalCents: 0 };
}

export async function checkAvailability(productId: string, variantId: string): Promise<number> {
  const { variant } = await resolveVariant(productId, variantId);
  return (variant as unknown as { stock: { available: number } }).stock.available;
}

File: java/src/main/java/com/ecommerce/backend/modules/cart/CartService.java
package com.ecommerce.backend.modules.cart;

import com.ecommerce.backend.common.error.ApiException;
import com.ecommerce.backend.modules.audit.AuditService;
import com.ecommerce.backend.modules.catalog.Product;
import com.ecommerce.backend.modules.catalog.Variant;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.bson.types.ObjectId;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;
import org.springframework.stereotype.Service;

/**
 * Cart business logic (mirrors {@code cart.service.ts}). The cart stores only
 * references and quantities — every price is re-read from the product document
 * at read time and the order module re-derives totals from the database at
 * checkout; client-supplied prices are never trusted.
 */
@Service
public class CartService {

    private final CartRepository cartRepository;
    private final MongoTemplate mongo;
    private final AuditService auditService;

    public CartService(CartRepository cartRepository, MongoTemplate mongo, AuditService auditService) {
        this.cartRepository = cartRepository;
        this.mongo = mongo;
        this.auditService = auditService;
    }

    // ------------------------------------------------------------- Public shape

    public record CartItemPublic(
            String productId,
            String variantId,
            String sku,
            String productName,
            String productSlug,
            String image,
            Map<String, String> attributes,
            long unitPriceCents,
            Long compareAtPriceCents,
            double taxRate,
            int quantity,
            long available,
            long lineTotalCents,
            boolean inStock,
            boolean productAvailable) {}

    public record CartPublic(List<CartItemPublic> items, int itemCount, long subtotalCents) {}

    // -------------------------------------------------------------- Operations

    public CartPublic getCart(String userId) {
        Cart cart = getOrCreateCart(userId);
        return enrichCart(cart.items);
    }

    public CartPublic addItem(String userId, String productId, String variantId, int quantity) {
        resolveVariant(productId, variantId);

        Cart cart = getOrCreateCart(userId);
        Cart.CartItem existing = findItem(cart, variantId);
        int newQuantity = (existing == null ? 0 : existing.quantity) + quantity;

        if (newQuantity > Cart.MAX_ITEM_QUANTITY) {
            throw ApiException.badRequest(
                    "Maximum of " + Cart.MAX_ITEM_QUANTITY + " units per item",
                    "QUANTITY_LIMIT_EXCEEDED");
        }

        int availability = checkAvailability(productId, variantId);
        if (newQuantity > availability) {
            throw ApiException.badRequest(
                    "Only " + availability + " units are currently available", "INSUFFICIENT_STOCK");
        }

        if (existing != null) {
            existing.quantity = newQuantity;
        } else {
            Cart.CartItem item = new Cart.CartItem();
            item.productId = productId;
            item.variantId = variantId;
            item.quantity = quantity;
            item.addedAt = Instant.now();
            cart.items.add(item);
        }
        save(cart);
        auditService.log("cart.item_added", userId, null,
                Map.of("productId", productId, "variantId", variantId, "quantity", quantity));
        return enrichCart(cart.items);
    }

    public CartPublic updateItemQuantity(String userId, String variantId, int quantity) {
        Cart cart = getOrCreateCart(userId);
        Cart.CartItem item = findItem(cart, variantId);
        if (item == null) {
            throw ApiException.notFound("Item not found in cart");
        }

        if (quantity > Cart.MAX_ITEM_QUANTITY) {
            throw ApiException.badRequest(
                    "Maximum of " + Cart.MAX_ITEM_QUANTITY + " units per item",
                    "QUANTITY_LIMIT_EXCEEDED");
        }

        int availability = checkAvailability(item.productId, variantId);
        if (quantity > availability) {
            throw ApiException.badRequest(
                    "Only " + availability + " units are currently available", "INSUFFICIENT_STOCK");
        }

        item.quantity = quantity;
        save(cart);
        auditService.log("cart.item_updated", userId, null,
                Map.of("variantId", variantId, "quantity", quantity));
        return enrichCart(cart.items);
    }

    public CartPublic removeItem(String userId, String variantId) {
        Cart cart = getOrCreateCart(userId);
        int before = cart.items.size();
        cart.items.removeIf(i -> variantId.equals(i.variantId));
        if (cart.items.size() == before) {
            throw ApiException.notFound("Item not found in cart");
        }
        save(cart);
        auditService.log("cart.item_removed", userId, null, Map.of("variantId", variantId));
        return enrichCart(cart.items);
    }

    public CartPublic clearCart(String userId) {
        Cart cart = getOrCreateCart(userId);
        cart.items = new ArrayList<>();
        save(cart);
        auditService.log("cart.cleared", userId, null, Map.of());
        return new CartPublic(List.of(), 0, 0);
    }

    // ------------------------------------------------------------- Internals

    public int checkAvailability(String productId, String variantId) {
        Resolved resolved = resolveVariant(productId, variantId);
        return Math.toIntExact(resolved.variant().available());
    }

    private Cart getOrCreateCart(String userId) {
        return cartRepository.findByUserId(userId).orElseGet(() -> {
            Cart cart = new Cart();
            cart.userId = userId;
            cart.items = new ArrayList<>();
            cart.createdAt = Instant.now();
            cart.updatedAt = cart.createdAt;
            return cartRepository.save(cart);
        });
    }

    private void save(Cart cart) {
        cart.updatedAt = Instant.now();
        cartRepository.save(cart);
    }

    private Cart.CartItem findItem(Cart cart, String variantId) {
        for (Cart.CartItem item : cart.items) {
            if (variantId.equals(item.variantId)) {
                return item;
            }
        }
        return null;
    }

    private record Resolved(Product product, Variant variant) {}

    private Resolved resolveVariant(String productId, String variantId) {
        if (!ObjectId.isValid(productId) || !ObjectId.isValid(variantId)) {
            throw ApiException.badRequest("Product variant does not exist", "VARIANT_NOT_FOUND");
        }
        Product product = mongo.findOne(
                Query.query(Criteria.where("_id").is(productId).and("variants._id").is(variantId)),
                Product.class);
        if (product == null) {
            throw ApiException.badRequest("Product variant does not exist", "VARIANT_NOT_FOUND");
        }
        if (!Product.STATUS_PUBLISHED.equals(product.status) || !product.isActive) {
            throw ApiException.badRequest(
                    "This product is not available for purchase", "PRODUCT_UNAVAILABLE");
        }
        Variant variant = findVariant(product, variantId);
        if (variant == null) {
            throw ApiException.badRequest("Product variant does not exist", "VARIANT_NOT_FOUND");
        }
        if (!variant.isActive) {
            throw ApiException.badRequest("This variant is not available", "VARIANT_UNAVAILABLE");
        }
        return new Resolved(product, variant);
    }

    private Variant findVariant(Product product, String variantId) {
        for (Variant variant : product.variants) {
            if (variantId.equals(variant.id)) {
                return variant;
            }
        }
        return null;
    }

    private CartPublic enrichCart(List<Cart.CartItem> items) {
        if (items.isEmpty()) {
            return new CartPublic(List.of(), 0, 0);
        }
        List<String> variantIds = items.stream().map(i -> i.variantId).toList();
        List<Product> products = mongo.find(
                Query.query(Criteria.where("variants._id").in(variantIds)), Product.class);

        List<CartItemPublic> publicItems = new ArrayList<>();
        long subtotalCents = 0;

        for (Cart.CartItem item : items) {
            Product product = products.stream()
                    .filter(p -> p.variants.stream().anyMatch(v -> item.variantId.equals(v.id)))
                    .findFirst()
                    .orElse(null);
            Variant variant = product == null ? null : findVariant(product, item.variantId);

            if (product == null || variant == null) {
                publicItems.add(new CartItemPublic(
                        item.productId,
                        item.variantId,
                        "",
                        "Unavailable product",
                        "",
                        null,
                        Map.of(),
                        0,
                        null,
                        0,
                        item.quantity,
                        0,
                        0,
                        false,
                        false));
                continue;
            }

            long unitPriceCents = variant.priceCents;
            long lineTotalCents = unitPriceCents * item.quantity;
            subtotalCents += lineTotalCents;

            String image = firstOf(product.images, variant.images);
            boolean productAvailable = Product.STATUS_PUBLISHED.equals(product.status)
                    && product.isActive
                    && variant.isActive;
            publicItems.add(new CartItemPublic(
                    item.productId,
                    item.variantId,
                    variant.sku,
                    product.name,
                    product.slug,
                    image,
                    variant.attributes == null ? new LinkedHashMap<>() : variant.attributes,
                    unitPriceCents,
                    variant.compareAtPriceCents,
                    variant.taxRate,
                    item.quantity,
                    variant.available(),
                    lineTotalCents,
                    variant.available() > 0,
                    productAvailable));
        }

        int itemCount = items.stream().mapToInt(i -> i.quantity).sum();
        return new CartPublic(publicItems, itemCount, subtotalCents);
    }

    private static String firstOf(List<String> first, List<String> second) {
        if (first != null && !first.isEmpty()) {
            return first.get(0);
        }
        if (second != null && !second.isEmpty()) {
            return second.get(0);
        }
        return null;
    }
}
